### PR TITLE
Add facilities for automatic token mapping

### PR DESCRIPTION
**Pull request by iwahbe**
This PR adds methods on `tfbridge.ProviderInfo` to apply naming strategies to missing tokens, as well as two common and simple strategies: single module and picking from a curated list of modules.

Token strategies are designed for an additive path from manual to automatic mapping, so there is no need to change any old code to start using automatic mapping. Strategies support an `Unmappable` method, which marks a substring as illegal to map, and thus for a human to map manually. This is for difficult to map or inconsistently mapped strings like `db => DB` or `flowlogs => FlowLogs`.

Example usage:
```go
// For a already defined `*tfbridge.ProviderInfo` called `info`
err := info.DefaultTokens(tfbridge.TokensSingleModule("pkgname_", "index", 
  func(module, name string) (string, error) {
    modulePostfix := lowerFirstChar(name)
    return fmt.Sprintf("pkgName:%s/%s:%s", module, modulePostfix, name), nil
  }))
```

### CODE REVIEW

**Comment by t0yv0** [comment]
How does this compare/work with the existing machinery?

- [TerraformToPulumiName](https://github.com/pulumi/pulumi-terraform-bridge/blob/6ffb2faf051e5faaf2234a0d0b926cba19c339ec/pkg/tfbridge/names.go#L83-#L83)

- [dataSourceName](https://github.com/pulumi/pulumi-terraform-bridge/blob/6ffb2faf051e5faaf2234a0d0b926cba19c339ec/pkg/tfgen/generate.go#L1529-#L1529)

- [resourceName](https://github.com/pulumi/pulumi-terraform-bridge/blob/6ffb2faf051e5faaf2234a0d0b926cba19c339ec/pkg/tfgen/generate.go#L1542-#L1542)

**Comment by t0yv0** [comment]
For resources, specifically, afaik if an tfbridge.ResourceInfo is missing/has no Tok specified, then tfgen will make up a module for that one resource and roll with it. Is the idea here to improve upon that to be able to semi-intelligently group resources into modules instead of just having an automatic module per resource?

**Comment by iwahbe** [comment]
As far as I can tell, the existing machinery exists to facilitate token generation from resources/datasources that are already mapped. This means that a manual mapping is still required, even if it is trivial, and that modules are not supported.

This PRs contribution works at the `tfbridge.ProviderInfo` level (just like manual mappings) and does not require human intervention to add new resources. Because the calling provider controls the `(module, name) -> token` mapping, it supports arbitrary token conventions (like `pkg:mod:Resource` instead of `pkg:mod/resource:Resource`) and modules. 

The strategies defined in this PR support mapping into pre-defined modules. Mapping into heuristically defined modules is a harder problem, requiring state between updates. I believe we can support heuristically defined modules in the future using this API, but it's not included in this PR.

**Comment by iwahbe** [comment]
> For resources, specifically, afaik if an tfbridge.ResourceInfo is missing/has no Tok specified, then tfgen will make up a module for that one resource and roll with it. Is the idea here to improve upon that to be able to semi-intelligently group resources into modules instead of just having an automatic module per resource?

To answer your question directly: kind of. The idea here is to obviate the need for further human intervention in the `resources.go` file for mappings that are (a) common and (b) machine computable. The obvious use case is slotting a resource/datasource into a pre-defined module based on its prefix (or slotting all resources into an `index` module).

File: pkg/tfbridge/info.go [hunk 1]
```diff
@@ -97,6 +99,62 @@ type ProviderInfo struct {
...
+func (info *ProviderInfo) ComputeDataSourceTokens(strategy DataSourceTokenStrategy) error {
+	if info.DataSources == nil {
+		info.DataSources = map[string]*DataSourceInfo{}
+	}
+	return applyComputedTokens(info.P.DataSourcesMap(), info.DataSources, strategy)
+}
+
+// For each key in the info map not present in the result map, compute a result and store
+// it in the result map.
+func applyComputedTokens[T ResourceInfo | DataSourceInfo](
```
Comment (t0yv0) [inline pkg/tfbridge/info.go:129]: TIL new Go feature here.

**Comment by t0yv0** [comment]
@iwahbe I see how this makes writing providers a little easier when we do not want to manually specify the mapping. I'm in favor of going forward with this but I'd like to request this API is marked experimental so users do not expect long term support and we have a chance to take the time to spit-ball some more alternatives before committing long-term to the API. Even if we don't change it I'd like to put some more eyes on it to refine documentation at least, it looks a little odd at a first glance (which is totally fine for the first draft)!

Also I just wanted to confirm a few assumptions so we're on the same page.

# Refining Assumptions


# API alternative

Imagining an alternative API. I understand that currently we have this which is limiting as the user thinks she needs to manually populate the map keys because it's not obvious she can write code to extract the terraform keys from shim.Provider:

```
ProviderInfo {
	P              shim.Provider // the TF provider/schema.
	Resources      map[string]*ResourceInfo // a map of TF name to Pulumi name; standard mangling occurs if no entry.
	IgnoreMappings []string                 // a list of TF resources and data sources to ignore in mappings errors
}
```

So an alternative API would lift `map[string]*ResourceInfo` to a function and provide some helper funcs accessible in that function context object. If this function is provided, the framework would tabulate the Resources map automatically by calling this function for every TF resource and reify ` map[string]*ResourceInfo` just as in your PR. 

It seems to me that assigning a Pulumi model to a resource is free-form, and the "strategies" do not need to be constrained beyond a `string->string` function, in which case modeling it as such is lower cognitive load than modeling it as strategy classes. I'd be curious if I'm missing something and the API needs more constraint.

Another advantage here is that the user is guided to do any further "automated" ResourceInfo customization as needed.


```
ProviderInfo {
        ResourceMapping func(ctx *ResourceMappingContext) error
  
        // Or this to emphasize input-output functional direction.
        ResourceMapping func(ctx *ResourceMappingContext, defaultResourceInfo ResourceInfo) (ResourceInfo, error)
}

// Extensible
type ResourceMappingContext {
     // Input, for example "random_string"
     TerraformResourceKey string

     // Output pre-populated with defaults.
     ResourceInfo ResourceInfo
}

func (ctx *ResourceMappingContext) SetModule(moduleName string) {}
func (ctx *ResourceMappingContext) Ignore() {}

resourceMapping := func (ctx *ResourceMappingContext) error {
   if strings.HasPrefix(ctx.TerraformResourceKey, "random_") {
     ctx.info.SetModule("index")
   }

   if strings.HasPrefix(ctx.TerraformResourceKey, "cs101_fizz_three") {
     ctx.info.SetModule("fizz")
   }

   if ... {
     // If we want to ingore the resource and not map it to Pulumi.
     ctx.Ignore()
   }

   return nil
}
```

**Comment by t0yv0** [comment]
Note for other reviewers: I've not found any sane ref documentation on Pulumi modules and the TF resource -> Pulumi module allocation that happens implicitly here in tfgen, therefore I'm relying somewhat on tribal knowledge here. If this is confusing - we can schedule a quick call to go over this and identify what we might want to document/clarify going forward.

File: pkg/tfbridge/token.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	return TokensKnownModules(tfPackagePrefix, moduleName, nil, finalize)
+}
+
+func tokensKnownModules[T ResourceInfo | DataSourceInfo](
+	prefix, defaultModule string, modules []string, new func(string, string) (*T, error),
+) TokenStrategy[T] {
+	return func(tfToken string, _ []string) (*T, error) {
+		tk := strings.TrimPrefix(tfToken, prefix)
+		if len(tk) == len(tfToken) {
+			return nil, fmt.Errorf("token '%s' missing package prefix '%s'", tfToken, prefix)
```
[the remark this answers is not in the document]
Comment (iwahbe) [reply pkg/tfbridge/token.go:68]: I know that `%q` exists, but it does something fundamentally different than `%s`. `%q` takes an arbitrary object and displays it with Go syntax. `%q` isn't robust to type changes or odd types ([example](https://go.dev/play/p/I6U5TefoMWs)). `%s` requires that the underlying type is either a `string` or implements `fmt.Stringer`, and is thus IMO safer to use. It also allows single quotes, which we use more often in our error messages.

**Comment by t0yv0** [comment]
👍

File: pkg/tfbridge/token.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lblackstone) [reply pkg/tfbridge/token.go:84]: TIL -- thanks for the example!

**Pull request merged**